File: lib/onr_login.py
"""Login ONR (LoginUsuarioCertificado) para obter tokens/Hash."""

from __future__ import annotations

from zeep import Client
from zeep.transports import Transport

from lib.cert_extract import extract_from_pfx
from lib.onr_env import resolve_path


def login_tokens(cfg: dict) -> list[str]:
    wsdl = resolve_path(cfg["wsdl_path"])
    if not wsdl.is_file():
        raise FileNotFoundError(f"WSDL de login não encontrado: {cfg['wsdl_path']}")

    cert_fields = extract_from_pfx(
        cfg["cert_path"],
        cfg["cert_password"],
        publickey_format=cfg.get("publickey_format", "base64_der"),
        validuntil_format=cfg.get("validuntil_format", "iso"),
    )

    o_request = {
        "SUBJECTCN": cert_fields["SUBJECTCN"],
        "ISSUERO": cert_fields["ISSUERO"],
        "PUBLICKEY": cert_fields["PUBLICKEY"],
        "SERIALNUMBER": cert_fields["SERIALNUMBER"],
        "VALIDUNTIL": cert_fields["VALIDUNTIL"],
        "CPF": cfg["cpf"],
        "EMAIL": cfg["email"],
        "IDParceiroWS": cfg["id_parceiro_ws"],
    }

    transport = Transport(timeout=60)
    client = Client(str(wsdl), transport=transport)
    client.service._binding_options["address"] = cfg["endpoint"]

    result = client.service.LoginUsuarioCertificado(oRequest=o_request)
    if not result.RETORNO:
        raise RuntimeError(
            f"Login falhou: [{result.CODIGOERRO}] {result.ERRODESCRICAO}"
        )

    if not result.Tokens:
        raise RuntimeError("Login OK, porém nenhum token (Hash) foi retornado.")

    tokens = result.Tokens.string
    return list(tokens) if isinstance(tokens, list) else [tokens]

File: lib/onr_login.js
import fs from "node:fs";
import soap from "soap";
import { extractFromPfx } from "./cert_extract.js";
import { resolvePath } from "./onr_env.js";

export async function loginTokens(cfg) {
  const wsdlPath = resolvePath(cfg.wsdlPath);
  if (!fs.existsSync(wsdlPath)) {
    throw new Error(`WSDL de login não encontrado: ${wsdlPath}`);
  }

  const certFields = extractFromPfx(cfg.certPath, cfg.certPassword, {
    publickeyFormat: cfg.publickeyFormat || "base64_der",
    validuntilFormat: cfg.validuntilFormat || "iso",
  });

  const oRequest = {
    SUBJECTCN: certFields.SUBJECTCN,
    ISSUERO: certFields.ISSUERO,
    PUBLICKEY: certFields.PUBLICKEY,
    SERIALNUMBER: certFields.SERIALNUMBER,
    VALIDUNTIL: certFields.VALIDUNTIL,
    CPF: cfg.cpf,
    EMAIL: cfg.email,
    IDParceiroWS: cfg.idParceiroWs,
  };

  const client = await soap.createClientAsync(wsdlPath, {
    endpoint: cfg.endpoint,
    forceSoap12Headers: false,
    wsdl_options: { timeout: 60000 },
  });

  const [response] = await client.LoginUsuarioCertificadoAsync({ oRequest });
  const result =
    response?.LoginUsuarioCertificadoResult ??
    response?.LoginUsuarioCertificadoResponse?.LoginUsuarioCertificadoResult ??
    response;

  if (!result?.RETORNO) {
    throw new Error(
      `Login falhou: [${result?.CODIGOERRO}] ${result?.ERRODESCRICAO}`
    );
  }

  const tokens = result?.Tokens?.string;
  const list = Array.isArray(tokens) ? tokens : tokens ? [tokens] : [];
  if (!list.length) {
    throw new Error("Login OK, porém nenhum token (Hash) foi retornado.");
  }
  return list;
}
